extmod/uasyncio.py

```diff
@@ -0,0 +1,565 @@
+                raise
+        # Lock available, set it as locked
+        self.state = 2
+        return True
+    async def __aenter__(self):
+        return await self.acquire()
+    async def __aexit__(self, exc_type, exc, tb):
+        return self.release()
+
+################################################################################
+# Event (optional component)
+
+# Event class for primitive events that can be waited on, set, and cleared
+class Event:
+    def __init__(self):
+        self.state = 0 # 0=unset; 1=set
+        self.waiting = Queue() # Queue of Tasks waiting on completion of this event
+    def set(self):
+        # Event becomes set, schedule any tasks waiting on it
+        while self.waiting.peek():
+            _queue.push_head(self.waiting.pop_head())
+        self.state = 1
+    def clear(self):
+        self.state = 0
+    async def wait(self):
+        if self.state == 0:
+            # Event not set, put the calling task on the event's waiting queue
+            self.waiting.push_head(cur_task)
+            # Set calling task's data to the event's queue so it can be removed if needed
+            cur_task.data = self.waiting
+            yield
+        return True
+
+################################################################################
+# General streams
+
+# Queue and poller for stream IO
+class IOQueue:
+    def __init__(self):
+        self.poller = select.poll()
+        self.map = {}
+    def _queue(self, s, idx):
+        if id(s) not in self.map:
+            entry = [None, None, s]
+            entry[idx] = cur_task
+            self.map[id(s)] = entry
+            self.poller.register(s, select.POLLIN if idx == 0 else select.POLLOUT)
+        else:
+            sm = self.map[id(s)]
+            assert sm[idx] is None
+            assert sm[1 - idx] is not None
+            sm[idx] = cur_task
+            self.poller.modify(s, select.POLLIN | select.POLLOUT)
+        # Link task to this IOQueue so it can be removed if needed
+        cur_task.data = self
+    def _dequeue(self, s):
+        del self.map[id(s)]
+        self.poller.unregister(s)
+    def queue_read(self, s):
+        self._queue(s, 0)
+    def queue_write(self, s):
+        self._queue(s, 1)
+    def remove(self, task):
+        while True:
+            del_s = None
+            for k in self.map: # Iterate without allocating on the heap
+                q0, q1, s = self.map[k]
+                if q0 is task or q1 is task:
+                    del_s = s
+                    break
+            if del_s is not None:
+                self._dequeue(s)
+            else:
+                break
+    def wait_io_event(self, dt):
+        for s, ev in self.poller.ipoll(dt):
+            sm = self.map[id(s)]
+            err = ev & ~(select.POLLIN | select.POLLOUT)
+            #print('poll', s, sm, ev, err)
+            if ev & select.POLLIN or (err and sm[0] is not None):
+                _queue.push_head(sm[0])
```

Comment:
This relies on sm[0] being not-None if a POLLIN event comes in. Does the implementation of ipoll guarantee that even in situations where an event is pending internally and then the Task in sm[0] is cancelled and removed before wait_io_event picks the event up? I guess them sm would be None... But maybe also refactoring into:
```
(ev & select.POLLIN or err) and sm[0] is not None
```
is more defensive and shouldn't increase code size. Same for POLLOUT.

Reply:
Good idea.  And then it further simplifies to `ev & ~select.POLLOUT and sm[0] is not None`